extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java

```diff
@@ -111,14 +112,23 @@ public QueryResponse<Object[]> runQuery(final DruidQuery druidQuery)
     Hook.QUERY_PLAN.run(druidQuery.getQuery());
     String taskId = MSQTasks.controllerTaskId(plannerContext.getSqlQueryId());
 
-    QueryContext queryContext = plannerContext.queryContext();
-    String msqMode = MultiStageQueryContext.getMSQMode(queryContext);
+    // SQL query context: context provided by the user, and potentially modified by handlers during planning.
+    // Does not directly influence task execution, but it does form the basis for the initial native query context,
+    // which *does* influence task execution.
+    final QueryContext sqlQueryContext = plannerContext.queryContext();
+
+    // Query context overrides: things that we add to the native query context prior to creating a controller task.
+    // These influence task execution.
+    final Map<String, Object> queryContextOverrides = new HashMap<>();
+
+    final String msqMode = MultiStageQueryContext.getMSQMode(sqlQueryContext);
     if (msqMode != null) {
-      MSQMode.populateDefaultQueryContext(msqMode, plannerContext.queryContextMap());
+      MSQMode.populateDefaultQueryContext(msqMode, queryContextOverrides);
```

Comment:
Lets say the user sets the mode as `strict` with `maxParseExceptions` as 100. 
Now what will happen is that `MSQMode.populateDefaultQueryContext` will populate `ImmutableMap.of(MSQWarnings.CTX_MAX_PARSE_EXCEPTIONS_ALLOWED, 0)` to the new `queryContextOverrides` map. 

The key `CTX_MAX_PARSE_EXCEPTIONS_ALLOWED` will be removed on line no:127 `queryContextOverrides.entrySet().removeIf(entry -> sqlQueryContext.containsKey(entry.getKey()));` since we already pass the `maxParseExceptions` in the original context. 

Hence I think we should throw a validation error here instead of removing it since we would not honor the strict ingestion setting.

Reply:
Hmm, good point. I was intending to preserve the old behavior. I adjusted this to make it so specific properties override the mode (as was the previous behavior) and added a test case to `MSQWarningsTest`.